packages/opentelemetry-plugin-postgres/src/pg.ts

```diff
@@ -0,0 +1,160 @@
+import { PostgresCallback, PostgresPluginOptions } from './types';
+import * as path from 'path';
+import * as pgTypes from 'pg';
+import * as shimmer from 'shimmer';
+
+export class PostgresPlugin extends BasePlugin<typeof pgTypes> {
+  static readonly component = 'pg';
+  readonly supportedVersions = ['^7.12.1'];
+  protected _config: PostgresPluginOptions;
+
+  constructor(readonly moduleName: string, readonly version: string) {
+    super();
+    this._config = {};
+  }
+
+  protected patch(): typeof pgTypes {
+    if (this._moduleExports.Client.prototype.query) {
+      shimmer.wrap(
+        this._moduleExports.Client.prototype,
+        'query',
+        this._getClientQueryPatch() as never
+      );
+    }
+    return this._moduleExports;
+  }
+  protected unpatch(): void {
+    if (this._moduleExports.Client.prototype.query) {
+      shimmer.unwrap(this._moduleExports.Client.prototype, 'query');
+    }
+  }
+
+  private _getClientQueryPatch() {
+    const plugin = this;
+    return (original: typeof pgTypes.Client.prototype.query) => {
+      plugin._logger.debug(
+        `Patching ${PostgresPlugin.component}.Client.prototype.query`
+      );
+      return function query(this: pgTypes.Client, ...args: unknown[]) {
+        // setup span
+        let callbackProvided: boolean =
+          args.length > 1 && typeof args[args.length - 1] === 'function';
+        const span = plugin._tracer.startSpan(
+          `${PostgresPlugin.component}.query`,
+          {
+            kind: SpanKind.CLIENT,
+            parent: plugin._tracer.getCurrentSpan() || undefined,
+            attributes: {
+              [AttributeNames.COMPONENT]: PostgresPlugin.component,
+              [AttributeNames.PG_HOST]: (this as any).connectionParameters.host,
+              [AttributeNames.PG_PORT]: (this as any).connectionParameters.port,
+            },
+          }
+        );
+
+        try {
+          if (typeof args[0] === 'string') {
+            span.setAttribute(AttributeNames.PG_TEXT, args[0]);
+            if (args[1] instanceof Array) {
+              span.setAttribute(AttributeNames.PG_VALUES, args[1]);
+              if (callbackProvided) {
+                args[2] = plugin._tracer.bind(args[2]);
+              }
+            } else {
+              if (callbackProvided) {
+                args[1] = plugin._tracer.bind(args[1]);
+              }
+            }
+          } else {
+            const config = args[0] as pgTypes.QueryConfig & {
+              callback?: PostgresCallback;
+            };
+            if (typeof config.name === 'string') {
+              span.setAttribute(AttributeNames.PG_PLAN, config.name);
+            } else {
+              if (typeof config.text === 'string') {
+                span.setAttribute(AttributeNames.PG_TEXT, config.text);
+              }
+              if (config.values instanceof Array) {
+                span.setAttribute(AttributeNames.PG_VALUES, config.values);
```

Comment:
Although `setAttribute` accepts `unknown` as a value, do you think we should flattern the array? Otherwise each exporterd needs to implement some logic to handle this.